File: leetcode/algorithms/3_longest_substring_without_repeating_characters/main.js
/**
 * Sliding Window (Using Array)
 *
 * Complexities:
 *   N - Length of `s`
 *   M - Length of Available Characters in `s`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(Min(M, N))
 */
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  const charMap = new Array(128).fill(-1);
  let left = 0;
  let maxLength = 0;

  for (let right = 0; right < s.length; right++) {
    const currentChar = s[right];

    if (charMap[currentChar] >= left) {
      left = charMap[currentChar] + 1;
    }

    charMap[currentChar] = right;

    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
};


// Solution
/**
 * Sliding Window (Using Map)
 *
 * Complexities:
 *   N - Length of `s`
 *   M - Length of Available Characters in `s`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(Min(M, N))
 */
var solution = function (s) {
  const map = new Map();
  let left = 0;
  let maxLength = 0;

  for (let right = 0; right < s.length; right++) {
    const char = s[right];

    if (map.has(char) && map.get(char) >= left) {
      left = map.get(char) + 1;
    }

    map.set(char, right);

    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
};

File: leetcode/algorithms/3_longest_substring_without_repeating_characters/main.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class LongestSubstringWithoutRepeatingCharacters {
    /**
     * Sliding Window
     *
     * Complexities:
     *   N - Length of `s`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public int lengthOfLongestSubstring(String s) {
        int[] charmap = new int[128];
        Arrays.fill(charmap, -1);

        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);

            if (charmap[currentChar] >= left) {
                left = charmap[currentChar] + 1;
            }

            charmap[currentChar] = right;

            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }


    // Solution
    /**
     * Sliding Window (using Map)
     *
     * Complexities:
     *   N - Length of `s`
     *   M - Length of Available Characters in `s`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(Min(M, N))
     */
    public int solution(String s) {
        Map<Character, Integer> map = new HashMap<>();

        int maxLength = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);

            if (map.containsKey(currentChar)) {
                left = Math.max(left, map.get(currentChar) + 1);
            }

            map.put(currentChar, right);

            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}
